React/CSS-in-JS/Chakra.md

```diff
@@ -0,0 +1,498 @@
+
+Example:
+
+```jsx
+import { ChakraProvider } from "@chakra-ui/react";
+
+function App({ Component, pageProps }: AppProps): ReactElement {
+  return (
+    <ChakraProvider theme={theme}>
+      <Component {...pageProps} />
+    </ChakraProvider>
+  );
+}
+
+export default App;
+```
+
+Note:
+
+- For Next.js, you need to set this up in `pages/_app.tsx`
+- For Create React App, you need to set this up in `index.tsx`
+
+### Style props
+
+Style props are a way to alter the style of a component by simply passing props to it. It helps to save time by providing helpful shorthand ways to style components.
+
+Alternative to that is using sx prop that takes an object with style rules.
+
+```tsx
+// shorthand width prop
+<Box w="400px">Lorem Ipsum is simply dummy text</Box>
+
+// sx prop
+<Box sx={{ width: '400px'}}>Lorem Ipsum is simply dummy text</Box>
+```
+
+For more about style props read [docs](https://chakra-ui.com/docs/features/style-props)
+
+#### The `as` prop
+
+The `as` prop is a feature that all Chakra UI components have and it allows you to pass an HTML tag or component to be rendered.
+
+For example, say you are using a `Button` component, and you need to make it a link instead. You can compose `a` and `Button` like this:
+
+```tsx
+<Button as="a" variant="outline" href="https://chakra-ui.com">
+  Lorem Ipsum
+</Button>
+```
+
+This allows you to use all of the `Button` props and all of the `a` props without having to wrap the `Button` in an `a` component.
+
+### Setup custom theme
+
+If you need to customize the default theme to match your design requirements, you can use `extendTheme` from `@chakra-ui/react`.
+
+Example:
+
+```jsx
+import { extendTheme } from "@chakra-ui/react";
+
+const overrides = {
+  colors: {
+    primary: {
+      100: "#ff0000",
+      80: "#ff1a1a",
+    },
+  },
+};
+
+const theme = extendTheme(overrides);
+
+export default theme;
+```
+
+For more info about setup read [docs](https://chakra-ui.com/docs/getting-started)
+
+## Theming rules
+
+The theme object is where you define your application's color palette, type scale, font stacks, breakpoints, border radius values, and more. Theme overrides are placed in `src/styles/themes` folder.
+
+Example:
+
+```ts
+const overrides = {
+  colors: {
+    primary: {
+      100: "#ff0000",
+      80: "#ff1a1a",
+    },
+  },
+};
+```
+
+Except objects, theme override can take a function:
+
+```ts
+const overrides = {
+  colors: (props) => {
+    const { colorMode } = props;
+
+    return {
+      primary: {
+        100: colorMode === "dark" ? "#ff0000" : "#ffffff",
+        80: "#ff1a1a",
+      },
+    };
+  },
+};
+```
+
+Then in `props` we have access button props, `theme` object and `colorMode`.
+
+Note:
+
+- `mode(lightMode, darkMode)(props)` function is the same as `colorMode === "dark" ? darkMode : lightMode`.
+- `mode` function is part of `chakra-ui/theme-tools` package.
```

Comment:
there is also `orient` https://github.com/chakra-ui/chakra-ui/blob/develop/packages/theme-tools/src/component.ts#L44 
Example: https://github.com/chakra-ui/chakra-ui/blob/e20efc1d8d42ce9d3c2280a5731d59e771de53c4/packages/theme/src/components/slider.ts#L8

And a lot of color transformation helpers https://github.com/chakra-ui/chakra-ui/blob/develop/packages/theme-tools/src/color.ts
`transparentize` is a cool one
Example: https://github.com/chakra-ui/chakra-ui/blob/abac2f999f3996ffc3ce575de07126dc9cc41bbb/packages/theme/src/components/badge.ts#L13